src/codegen.cpp

```diff
@@ -2564,8 +2564,8 @@ static Value *emit_f_is(jl_codectx_t &ctx, const jl_cgval_t &arg1, const jl_cgva
         (jl_pointer_egal(rt1) || jl_pointer_egal(rt2)))
         return ctx.builder.CreateICmpEQ(boxed(ctx, arg1), boxed(ctx, arg2));
 
-    bool justbits1 = jl_is_concrete_immutable(rt1);
-    bool justbits2 = jl_is_concrete_immutable(rt2);
+    bool justbits1 = jl_is_concrete_immutable(rt1) && ((jl_datatype_t*)rt1)->isinlinealloc;
+    bool justbits2 = jl_is_concrete_immutable(rt2) && ((jl_datatype_t*)rt2)->isinlinealloc;
```

Comment:
The old implementation avoids this by calling `emit_box_compare` for pointer fields so I think a better fix to make sure there's no regression is to do the check in `emit_bits_compare` on `fldty` instead when the field is a pointer.